drivers/usb/bc12/bc12_pi3usb9201.c

```diff
@@ -0,0 +1,392 @@
+	 * the callbacks to clear out the BC1.2 state.
+	 */
+	if (pi3usb9201_data->partner_type != BC12_TYPE_NONE) {
+		pi3usb9201_notify_callback(dev, NULL);
+	}
+
+	if (state) {
+		/* Now update the current charger type */
+		pi3usb9201_data->partner_type = state->type;
+		/* If new type != NONE, then notify charge manager */
+		if (state->type != BC12_TYPE_NONE) {
+			pi3usb9201_notify_callback(dev, state);
+		}
+	} else {
+		pi3usb9201_data->partner_type = BC12_TYPE_NONE;
+	}
+}
+
+static int pi3usb9201_client_detect_start(const struct device *dev)
+{
+	int rv;
+
+	/*
+	 * Read both status registers to ensure that all interrupt indications
+	 * are cleared prior to starting bc1.2 detection.
+	 */
+	pi3usb9201_get_status(dev, NULL, NULL);
+
+	/* Put pi3usb9201 into client mode */
+	rv = pi3usb9201_set_mode(dev, PI3USB9201_CLIENT_MODE);
+	if (rv)
+		return rv;
+	/* Have pi3usb9201 start bc1.2 detection */
+	rv = pi3usb9201_bc12_detect_ctrl(dev, true);
+	if (rv)
+		return rv;
+	/* Enable interrupt to wake task when detection completes */
+	return pi3usb9201_interrupt_enable(dev, true);
+}
+
+static void pi3usb9201_client_detect_finish(const struct device *dev, int client_status)
+{
+	struct bc12_partner_state new_chg;
+	int bit_pos;
+
+	/* Set charge voltage to 5V */
+	new_chg.voltage = BC12_CHARGER_VOLTAGE_MV;
+
+	/*
+	 * Find set bit position. Note that this function is only called if a
+	 * bit was set in client_status, so bit_pos won't be negative.
+	 */
+	bit_pos = __builtin_ffs(client_status) - 1;
+
+	new_chg.current = bc12_chg_limits[bit_pos].current_limit;
+	new_chg.type = bc12_chg_limits[bit_pos].partner_type;
+
+	LOG_DBG("client_sts = 0x%x, current = %d mA, type = %d", client_status, new_chg.current,
+		new_chg.type);
+
+	/* bc1.2 is complete and start bit does not auto clear */
+	pi3usb9201_bc12_detect_ctrl(dev, false);
+	/* Inform charge manager of new supplier type and current limit */
+	pi3usb9201_update_charging_partner(dev, &new_chg);
+}
+
+static int pi3usb9201_disconnect(const struct device *dev)
+{
+	/* Put pi3usb9201 into its power down mode */
+	pi3usb9201_set_mode(dev, PI3USB9201_POWER_DOWN);
+	/* The start bc1.2 bit does not auto clear */
+	pi3usb9201_bc12_detect_ctrl(dev, false);
+	/* Mask interrupts unitl next bc1.2 detection event */
+	pi3usb9201_interrupt_enable(dev, false);
+
+	/*
+	 * Let the application know there's no more charge available for the
+	 * supplier type that was most recently detected.
+	 */
+	pi3usb9201_update_charging_partner(dev, NULL);
+
+	return 0;
+}
+
+static int pi3usb9201_set_portable_device(const struct device *dev)
+{
+	/* Disable interrupts during mode change */
+	pi3usb9201_interrupt_enable(dev, false);
+
+	if (pi3usb9201_client_detect_start(dev)) {
+		struct bc12_partner_state new_result;
+
+		/*
+		 * VBUS is present, but starting bc1.2 detection failed
+		 * for some reason. Set the partner type to SDP to disable
+		 * USB data signals and limit charing to 500 mA.
```

Comment:
Good point.  I guess in that case it will be up to the application code to determine when the device is configured and can override the BC1.2 reported current.

Do you see value in creating a new charging_partner type of unknown if BC1.2 detection fails?
```
enum bc12_type {
	BC12_TYPE_NONE,
	BC12_TYPE_SDP,
	BC12_TYPE_DCP,
	BC12_TYPE_CDP,
	BC12_TYPE_PROPRIETARY,
	BC12_TYPE_UNKNOWN,    
};
```

Reply:
I don't really understand what the "disable USB data signals" here refers to? How are the signals disabled? I don't quite see how the code would set Power Down mode in Control 1 register in this code path.

If the detection fails, then I think the current should be limited to 2.5 mA averaged over 1 second (I_SUSP). Then if you are connected to SDP (Standard Downstream Port, e.g. some host that does not support battery charging) then the current should be limited to 2.5 mA when suspended (no SOF on Full-/High-Speed and no keep-alive on Low-Speed), 100 mA when not configured and 500 mA when configured. Then there are 150 mA/900 mA for Super-Speed but atleast now Zephyr does not support these.

> Without knowing if the device is configured or not, must it not be limited to 100mA?

The device can draw up to 1.5 A if connected to CDP or DCP regardless if it is configured or not (on DCP the device won't ever be configured though as it has D+/- shorted).

Reply:
> > Without knowing if the device is configured or not, must it not be limited to 100mA?
> 

> The device can draw up to 1.5 A if connected to CDP or DCP regardless if it is configured or not (on DCP the device won't ever be configured though as it has D+/- shorted).

Fine, but it was about "partner type SDP". `"Set the partner type to SDP to disable USB data signals and limit charing to 500 mA."`

Reply:
The BC1.2 driver and chip doesn't have any information about when the USB port is suspended and configured.  This is up to the application to determine.  Furthermore, the BC1.2 driver should only disable the USB data signals when DCP is detected.  For other modes, the BC1.2 driver and application should assume the data path is available.

So I think for failed BC1.2 detection, the driver should do the following:
* Report the partner type as BC12_TYPE_UNKNOWN.
* Report the partner voltage as 5000 mV
* Report the partner current as 500mA
* Leave USB data signals on.

When the application receives the callback with BC12_TYPE_UNKNOWN or BC12_TYPE_SDP, it is the application's responsibility to determine the suspended/unconfigured/configured state, and set the charging limits appropriately (2.5 mA, 100 mA, or 500 mA).

Reply:
> Good point. I guess in that case it will be up to the application code to determine when the device is configured and can override the BC1.2 reported current.
> 
> Do you see value in creating a new charging_partner type of unknown if BC1.2 detection fails?

I think BC12_TYPE_UNKNOWN makes sense, but also set current_ma to 0 or ISUSP for BC12_TYPE_UNKNOWN/BC12_TYPE_SDP charging partner, and document that in these cases the charging manager should select the values based on USB device state.

Reply:
Updated both SDP and UNKNOWN modes to return 2.5 mA.